test(fixer): cover more test cases

jason89521 opened the pull request:
## Description

Almost all ESLint's test cases are ported. However, there are some test case cannot be satisfied:

1. BOM removal: ESLint use `range.start===-1` to determine whether to delete the starting BOM, but `Span` cannot accept negative number.
2. Messages sorting: If there are multiple nofix messages, eslint we sort them by there column and line.

## Related issue

#68

jason89521 commented:
For BOM removal, we can add a flag property in `Fix`. Is this approach appropriate?

jason89521 commented:
For messages sorting, we can add `Span` to `Message` struct.

Boshen commented:
Both are interesting problems, let me think about these.

Boshen commented:
For BOM, it seems our parser will parse it and offset all the spans correctly, can you test this and see if this is correct?

Boshen commented:
What is message sorting exactly? Can you give an example?

jason89521 commented:
> For BOM, it seems our parser will parse it and offset all the spans correctly, can you test this and see if this is correct?

I prepend a `\uFEFF` to the test.js, and the parser said that it is an invalid character. Not sure whether the way to check this is correct or not.
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/47748132/223056317-e688b2df-81a2-4456-be61-418f0d61934c.png">

jason89521 commented:
> What is message sorting exactly? Can you give an example?

Suppose we have two messages:

```
NoFix1 {
  error: no_fix_error1,
  line: 1,
  column: 0
  fix: None,
  fixed: false,
}

NoFix2 {
  error: no_fix_error2,
  line: 1,
  column: 10
  fix: None,
  fixed: false
}
```

Since they cannot be fixed, the fixer will consider them remaining unfixed message. For ESLint, they will sort them by the `line` and `column`. In this example, `NoFix1` will be in front of `NoFix2`.

Boshen commented:
```
oxc  main ❯ cargo run -p oxc_parser --example parser -- ./tasks/coverage/typescript/tests/cases/compiler/bom-utf8.ts
    Finished dev [unoptimized + debuginfo] target(s) in 0.35s
     Running `target/debug/examples/parser ./tasks/coverage/typescript/tests/cases/compiler/bom-utf8.ts`
{
  "type": "Program",
  "start": 0,
  "end": 14,
  "sourceType": "module",
  "body": [
    {
      "type": "VariableDeclaration",
      "start": 3, <-------------------------------- starts at 3
      "end": 12,
      "kind": "var",
      "declarations": [
```

I think eslint had to do this to be compatible with different parsers, which we don't need to.

Boshen commented:
> For ESLint, they will sort them by the line and column. In this example, NoFix1 will be in front of NoFix2.

Does this mean all we have to do is sort by span.start?

Boshen commented:
Why does it need to sort? For printing unfixed messages in order?

jason89521 commented:
I don't kwon why eslint sort the unfixed messages, I think it is a bit unnecessary. How do you think?

jason89521 commented:
Does the order of the unfixed message really matter?

Boshen commented:
> 

It's more readable and user friendly.

jason89521 commented:
> So the conclusion here is that we can drop the BOM test cases? 

Ya, let me remove them.

Boshen commented:
Can you write some comments on this in the place where you deleted the tests so the next person knows BOM handling is not implemented on purpose.

Boshen commented on crates/oxc_linter/src/autofix/fixer.rs:
```diff
@@ -356,170 +356,9 @@ mod test {
         ]);
         assert_eq!(result.fixed_code, TEST_CODE.replace("answer", "foo"));
         assert_eq!(result.messages.len(), 2);
+        // We should sort the error to pass the following assertion.
```
What does this comment mean?

Boshen replied:
Can you ignore this if it's not passing write now.

Boshen replied:
Oh, you should add a `TODO:` here.

jason89521 replied:
It is already ignored. It means that we need to sort the remaining unfixed messages.

Boshen replied:
Yes if it's easy to do so, let's do another PR for this after this.
